internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -90,33 +90,29 @@ object ClassToAPI {
    *
    * In order to fully fix this issue, the Scala compiler should do the desired behaviour
    * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
-   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   * issue is here: https://github.com/scala/bug/issues/2034.
    *
-   * @return The canonical name if not null, the full name otherwise.
+   * @return The canonical name if not null, the blank string otherwise.
    */
-  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+  def classCanonicalName(c: Class[_]): String = {
     if (c == null) "" // Return nothing if it hits the top-level class
     else {
       val className = c.getName
-      // Adds a `dollar` if it's a recursive call, else nothing
-      val atEnd: String = if (isRecursive) "$" else ""
       try {
         val canonicalName = c.getCanonicalName
         if (canonicalName == null) className
-        else canonicalName + atEnd
+        else canonicalName
       } catch {
         case malformedError: java.lang.InternalError
             if malformedError.getMessage.contains("Malformed class name") =>
           val enclosingClass = c.getEnclosingClass
           val enclosingName = enclosingClass.getName
           val restOfName = c.getName.stripPrefix(enclosingName)
-          handleMalformedNameOf(enclosingClass, true) + restOfName + atEnd
+          classCanonicalName(enclosingClass).stripSuffix("$") + "." + restOfName
```

Comment:
Then your concern is of the status quo, not specifically this proposed change. Because currently (as demonstrated in the spec as of #380) you can't tell the difference between:

+ OO and CO: both "x.y$"
+ OC and CC: both "x.y"
+ OCO and CCO: both "x.y.z$"
+ OCC and CCC: both "x.y.z"

and btw, these are coming direct from getCanonicalClass.

If Zinc depends on the invariant that distinct classes have distinct names (as returned by classCanonicalName) then there's more work to do here, starting with not depending on native getCanonicalClass.

Shall I open a ticket?